### ISSUE TITLE
MSAN does not always detect accessing uninitialized bytes after variant destroy

### ISSUE THREAD

**Issue by ccotter** [llvm/llvm-project#191883]
```c++
std::variant<double, int> v;
v.emplace<double>();
double& d = std::get<double>(v);
v.emplace<int>();
if (d) ...
```

MSAN does not catch the above program, when the type being destroyed due to the second replace is trivial / does not have its own destructor.

See https://github.com/llvm/llvm-project/pull/101048 for background. The fix suggested by this PR (now closed, since that approach is not the right approach) is to emit a destructor call for the trivial case, and have the MSAN runtime properly poison the destructed bytes.
